### ISSUE TITLE
freeze with video-sync=display-resample and vulkan-swap-mode=fifo

### ISSUE THREAD

**Issue by e00E** [mpv-player/mpv#7325]
### mpv version and platform
mpv master, arch linux, wayland, sway master
(also with current releases of mpv, sway)

My graphics card is a Radeon Rx 5700 XT. I am using the amdgpu driver and the amdvlk vulkan package.

When I use vulkan-radeon (radv) instead of amdvlk the problem does not occur but I get this message from mpv (not sure if mpv writes it or some library)

> WARNING: radv is not a conformant vulkan implementation, testing use only.

which is what initially prompted me to use amdvlk. This might indicate that it is a bug in amdvlk and not mpv. If you think so, I can report there but I expect they will want easier reproduction steps. Maybe someone has an idea what the problem is. It could still be a bug in mpv that just happens to not be triggered with radv.

### Reproduction steps
Play any video with `video-sync=display-resample`, `gpu-context=waylandvk` and `vulkan-swap-mode=fifo` or `vulkan-swap-mode=fifo-relaxed`.

The reason I am setting these is because the documentation states

> These modes also require a vsync blocked presentation mode. For OpenGL, this translates to `--opengl-swapinterval=1`. For Vulkan, it translates to `--vulkan-swap-mode=fifo` (or `fifo-relaxed`).

### Expected behavior
Video plays properly.

### Actual behavior
This happens in maybe 3 out of 4 attempts:

The mpv display window opens but freezes on one of the first few frames. The terminal status line displays time 00:00:00 but I can hear the audio play for a few seconds.
If I continuously move the mouse cursor within the mpv display window then the video plays correctly. Stopping the movement freezes the video again.
I can cause the video to update and freeze again by pausing and unpausing or adjusting the volume with the hotkey.

In the other 1 out of 4 attempts the video plays correctly.

When I don't specify `vulkan-swap-mode` or when I manually set it to `mailbox` or `immediate` then the problem does not occur. This is weird because it means that mpv automatically selects one of those modes even though they are not vsync blocked even though this is required for `video-sync=display-resample`. This might be a separate bug. The log file does not seem to contain any information on which `vulkan-swap-mode` is selected automatically and why.

### Log file
[mailbox.txt](https://github.com/mpv-player/mpv/files/4022657/mailbox.txt)
[fifo_without_bug.txt](https://github.com/mpv-player/mpv/files/4022658/fifo_without_bug.txt)
[fifo_with_bug.txt](https://github.com/mpv-player/mpv/files/4022659/fifo_with_bug.txt)

Locally I diffed the logs (remove the timestamps with a macro first) but did not see anything relevant.

**Comment by Dudemanguy** [comment]
Yeah this is a Wayland quirk that you either love or hate (I guess we hate it since we have to work around it). The swap buffers function on wayland actually blocks by default (i.e. it waits on vblank). This means when a window is hidden; rendering freezes until it is in view again. This has tons of problems for us so Wayland is forcibly set to mailbox in the backend (on opengl; the swapinterval is set to 0 which is the same thing) so it doesn't block. Instead there's a hack that basically waits for wayland events to appear on the queue with a timeout and uses that for vsync.

TL;DR: Wayland is only going to behave nicely with `mailbox` or other equivalent settings.

**Comment by e00E** [comment]
Thank you for the explanation. Just to be clear, in this case the window is not hidden, it is always visible. However I need to move the mouse within the window (as opposed to be a differente window) to make the video play.
I am surprised that it sometimes does work because your explanation makes it seem like it should never work with `fifo`.

**Comment by Dudemanguy** [comment]
>in this case the window is not hidden

Oh I misunderstood then sorry. 

Actually I think amdvlk is broken for your driver. The behavior you described doesn't make sense and I couldn't reproduce it here (RX 550) on amdvlk. Selecting fifo means rendering will be blocked on vsync (whatever the driver is doing) and the internal `vo_wayland_wait_frame`. If you use bump up the rendering settings (high quality shaders or whatever), you should notice that fifo will start dropping frames before mailbox does which makes sense because each frame is waiting on both the driver and mpv's `vo_wayland_wait_frame`. The reason why moving your mouse around would work is because it has to render the mouse which I guess amdvlk catches in the swap buffers call. It should normally catch mpv's rendering anyway though.

**Comment by haasn** [comment]
It's a bug in documentation. Setting mailbox is correct for wayland.
